### DISCUSSION TITLE
Force vcpkg toolchain to add full paths instead of relative paths

### DISCUSSION THREAD

**Discussion by netcorefan1** [microsoft/vcpkg#30559] [Q&A]
Hello,
I don't know why but for some fields vcpkg toolchain adds relative paths instead of full paths:
```
GEOTIFF:/include
GEOTIFF_LIRARY:C:/dev/vcpkg/installed/x64-windows/lib/geotiff_i.lib
```
This cause compilation failure because it cannot find GEOTIFF include directory.
It compiles only when I manually change path from `/include` into `C:/dev/vcpkg/installed/x64-windows/include`
How can I force vcpkg to use full paths? And why this is not done by default given that cmake doesn't know anything about installed packages and expect full paths?

**Comment by dg0yt** [comment]
Which "fields" do refer to? How do you get libgeotiff configuration?
(Is the problem really related to vcpkg, or to CMake, or to pkg-config?)

**Reply by netcorefan1** [reply]
I refer to the CMake variables to build OpenCv.
Without vcpkg toolchain:
```
-- Building zlib from source
CMake Warning (dev) at ../FindPackageHandleStandardArgs.cmake:438 (message):
  The package name passed to `find_package_handle_standard_args` (GEOTIFF)
  does not match the name of the calling package (GeoTIFF).  This can lead to
  problems in calling code that expects `find_package` result variables
  (e.g., `_FOUND`) to follow a certain pattern.
Call Stack (most recent call first):
  Emgu.CV.Extern/libgeotiff/cmake/FindGeoTIFF.cmake:81 (FIND_PACKAGE_HANDLE_STANDARD_ARGS)
  Emgu.CV.Extern/CMakeLists.txt:3852 (FIND_PACKAGE)
This warning is for project developers.  Use -Wno-dev to suppress it.

-- Could NOT find GEOTIFF (missing: GEOTIFF_LIBRARY)
-- CVEXTERN: GEOTIFF not found. Building from source
CMake Deprecation Warning at Emgu.CV.Extern/libgeotiff/CMakeLists.txt:33 (CMAKE_MINIMUM_REQUIRED):
  Compatibility with CMake < 2.8.12 will be removed from a future version of
  CMake.

  Update the VERSION argument <min> value or use a ...<max> suffix to tell
  CMake that the project does not need compatibility with older versions.

-- Enable in-code GeoTIFF EPSG tables: FALSE
-- Could NOT find PROJ4 (missing: PROJ4_LIBRARY PROJ4_INCLUDE_DIR)
-- Found ZLIB: zlib (found version "1.2.13")
-- Found JPEG: libjpeg-turbo
-- Generating geo_config.h header
-- Generating geo_config.h header - done
GEOTIFF_BIN_SUBDIR:STRING=bin
GEOTIFF_BUILD_PEDANTIC:BOOL=FALSE
GEOTIFF_DATA_SUBDIR:STRING=.
GEOTIFF_ENABLE_INCODE_EPSG:BOOL=FALSE
GEOTIFF_INCLUDE_DIR:STRING=/include
GEOTIFF_INCLUDE_SUBDIR:STRING=include
GEOTIFF_LIBRARY:FILEPATH=GEOTIFF_LIBRARY-NOTFOUND
GEOTIFF_LIB_SUBDIR:STRING=lib
```
With it:
```
-- Building zlib from source
CMake Warning (dev) at C:/Program Files/Microsoft Visual Studio/2022/Professional/Common7/IDE/CommonExtensions/Microsoft/CMake/CMake/share/cmake-3.25/Modules/FindPackageHandleStandardArgs.cmake:438 (message):
  The package name passed to `find_package_handle_standard_args` (GEOTIFF)
  does not match the name of the calling package (GeoTIFF).  This can lead to
  problems in calling code that expects `find_package` result variables
  (e.g., `_FOUND`) to follow a certain pattern.
Call Stack (most recent call first):
  Emgu.CV.Extern/libgeotiff/cmake/FindGeoTIFF.cmake:81 (FIND_PACKAGE_HANDLE_STANDARD_ARGS)
  C:/dev/vcpkg/scripts/buildsystems/vcpkg.cmake:852 (_find_package)
  Emgu.CV.Extern/CMakeLists.txt:3852 (FIND_PACKAGE)
This warning is for project developers.  Use -Wno-dev to suppress it.

-- Found GEOTIFF: C:/dev/vcpkg/installed/x64-windows/lib/geotiff_i.lib
-- CVEXTERN: GEOTIFF found. INCLUDE DIR: /include
GEOTIFF_INCLUDE_DIR:STRING=/include
GEOTIFF_LIBRARY:FILEPATH=C:/dev/vcpkg/installed/x64-windows/lib/geotiff_i.lib
(all the other variables are missing)
```
`FindGeoTIFF.cmake` content:
```
SET(GEOTIFF_NAMES geotiff)

IF(WIN32)

    IF(MINGW)
        FIND_PATH(GEOTIFF_INCLUDE_DIR
            geotiff.h
            PATH_PREFIXES geotiff
            PATHS
            /usr/local/include
            /usr/include
            c:/msys/local/include)

        FIND_LIBRARY(GEOTIFF_LIBRARY
            NAMES ${GEOTIFF_NAMES}
            PATHS
            /usr/local/lib
            /usr/lib
            c:/msys/local/lib)
    ENDIF(MINGW)

    IF(MSVC)
        SET(GEOTIFF_INCLUDE_DIR "$ENV{LIB_DIR}/include" CACHE STRING INTERNAL)

        SET(GEOTIFF_NAMES ${GEOTIFF_NAMES} geotiff_i)
        FIND_LIBRARY(GEOTIFF_LIBRARY NAMES 
            NAMES ${GEOTIFF_NAMES}
            PATHS
            "$ENV{LIB_DIR}/lib"
            /usr/lib
            c:/msys/local/lib)
    ENDIF(MSVC)
  
ELSEIF(UNIX)

    FIND_PATH(
      GEOTIFF_INCLUDE_DIR geotiff.h 
#      PATH_PREFIXES geotiff
      PATHS
        /usr/include/geotiff
        /usr/include/libgeotiff
)

    FIND_LIBRARY(GEOTIFF_LIBRARY NAMES ${GEOTIFF_NAMES})

ELSE()
    MESSAGE("FindGeoTIFF.cmake: unrecognized or unsupported operating system")
ENDIF()

IF(GEOTIFF_FOUND)
  SET(GEOTIFF_LIBRARIES ${GEOTIFF_LIBRARY})
ENDIF()

# Handle the QUIETLY and REQUIRED arguments and set SPATIALINDEX_FOUND to TRUE
# if all listed variables are TRUE
INCLUDE(FindPackageHandleStandardArgs)
FIND_PACKAGE_HANDLE_STANDARD_ARGS(GEOTIFF DEFAULT_MSG GEOTIFF_LIBRARY GEOTIFF_INCLUDE_DIR)
```
The relative `/include` has been set from `Emgu.CV.Extern`  which expects to build and link its own version of geotiff contained inside the project. I'm afraid that I'm encountering the same problem reported [here](https://github.com/microsoft/vcpkg/discussions/30497). The build system expect to build and link its own version of the library, but vcpkg takes over and I end up with a fatal mix (`include` coming from the build system and `geotiff_i.lib` from vcpkg). 
If I can manage on how to correctly hande such conflicts through `--overlay-ports` as you [suggested](https://github.com/microsoft/vcpkg/discussions/30497), then this and other issues should not represent a problem anymore.

**Reply by dg0yt** [reply]
- Vcpkg doesn't want vendored dependencies aka "own version of the library", for good reasons.
- `/include` isn't relative IMO. Here, it is the result of 
  ~~~cmake
  SET(GEOTIFF_INCLUDE_DIR "$ENV{LIB_DIR}/include" CACHE STRING INTERNAL)
  ~~~
  with empty environment variable `LIB_DIR`. You must provide this environment variable: By `INTERNAL` the constructed value is forced over user provided initialization. It is a poor find module... (EDIT : it should do `find_path(GEOTIFF_INCLUDE_DIR ...)` as it does `find_library(GEOTIFF_LIBRARY ...)`. This would relax the situation at least for dynamic linkage. Not being able to reliably handle static linkage with find modules is one reason for seeing vendored copies...)
- FTR libgeotiff in vcpkg has CMake config. It may not set those variables but it would probably be easy to leverage the config in the find module.

**Reply by netcorefan1** [reply]
Yes, and I completely agree with that policy. Most of the troubles I encounter comes from customized versions of the libraries. I tried so many times to replace them with vcpkg versions, but the projects does not build or give runtime errors.

Ohh...now I see!

Well, at this point all I need to do is to fix the above initialization and add a geotiff empty overlay port, is that right?

**Reply by dg0yt** [reply]
TBH I would suggest to use vcpkg's libgeotiff. (EDIT: Rest of comment integrated into previous comment.)

**Reply by netcorefan1** [reply]
Don't want to further abuse your time and I'm going to close the discussion as resolved. 
I would like just one last recommendation which will save me from bad headaches in the future.
You recommended to use vcpkg's libgeotiff and I will do for this and other libraries. However I must be prepared for when this is not possible due to the project requiring its own version.

Now, when the project fails to compile, I already know that I can't replace with a vcpkg version of the library.
But when succeeds, it can still generate runtime errors at unpredictable times. It can occur after one month of usage and only when I access a specific invalid memory section. It will be very hard at the point to find out that the error has been generated from having linked an incompatible version of the library rather than a generic error.
If a repo contains a submodule which has not been updated from very long time (very common), I can expect should be safe to replace it with the updated vcpkg version as long as there are no custom changes that affects its behavior and as long as there is not wrapper behind (which may lack the new function signatures).

My question is, how I can safely replace a library with vcpkg version without going into the (very times consuming) process of checking the whole history of changes made by the author (commits, wrapper signatures etc) ?

**_EDIT_**: It just happened exactly what was scaring me. I linked vcpkg libgeotiff. Project compiled well, the first function call returned a valid pointer, then with another different function: 
`System.AccessViolationException: 'Attempted to read or write protected memory. This is often an indication that other memory is corrupt.'`
With the original libgeotiff provided by the project, the same function worked as expected.
I have been lucky. I could had very well called another function instead of the bad one and end up with a false sense of safety.
I really need to find out some sort of security checks in order to link vcpkg packages safely. It's better to stay with an old and less featured library if there are even minimal risks like these.
